Library is unusable when only `html-strict` or `html-lenient` feature is enabled

Calling `Soup::html_strict()` with only the `html-strict` feature enabled and all other features disabled results in the following compilation error:

```
    Checking soupy v0.8.3
error[E0432]: unresolved import `crate::parser::html::HTMLNode`
  --> /Users/jwodder/.cargo/registry/src/index.crates.io-6f17d22bba15001f/soupy-0.8.3/src/parser/html/strict.rs:30:5
   |
30 | use crate::parser::html::HTMLNode;
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `HTMLNode` in `parser::html`
   |
note: found an item that was configured out
  --> /Users/jwodder/.cargo/registry/src/index.crates.io-6f17d22bba15001f/soupy-0.8.3/src/parser/html/mod.rs:11:15
   |
11 | pub use node::HTMLNode;
   |               ^^^^^^^^
note: the item is gated behind the `html` feature
  --> /Users/jwodder/.cargo/registry/src/index.crates.io-6f17d22bba15001f/soupy-0.8.3/src/parser/html/mod.rs:10:7
   |
10 | #[cfg(feature = "html")]
   |       ^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0432`.
error: could not compile `soupy` (lib) due to 1 previous error
```

The same thing happens when calling `Soup::html()` with only the `html-lenient` feature enabled.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -19,7 +19,7 @@ xml = ["dep:xmltree"]
 
 [dependencies]
 nom = { version = "7.1", optional = true }
-regex = { version = "1.9", optional = true }
-scraper = { version = "0.19", optional = true }
-ego-tree = { version = "0.6", optional = true }
-xmltree = { version = "0.10", optional = true }
+regex = { version = "1.11", optional = true }
+scraper = { version = "0.22", optional = true }
+ego-tree = { version = "0.10", optional = true }
+xmltree = { version = "0.11", optional = true }

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "soupy"
-version = "0.8.3"
+version = "0.9.0"
 edition = "2021"
 description = "A library for querying tree-based formats, similar to BeautifulSoup."
 keywords = ["beautifulsoup", "bs4", "soup", "html", "xml"]

--- a/src/parser/html/lenient.rs
+++ b/src/parser/html/lenient.rs
@@ -66,7 +66,10 @@ impl<'a> TryFrom<ego_tree::NodeRef<'a, scraper::Node>> for HTMLNode<scraper::Str
 
 #[cfg(test)]
 mod tests {
-    use crate::*;
+    use crate::{
+        parser::HTMLNode,
+        *,
+    };
 
     const HELLO: &str = r#"
 <!DOCTYPE html>
@@ -138,5 +141,99 @@ mod tests {
             Some("h3".into())
         );
         assert_eq!(headings.next().and_then(|h| h.name().cloned()), None);
+    }
+
+    #[test]
+    fn test_whitespace() {
+        let soup = Soup::html("<a> </a>");
+        assert_eq!(soup.nodes, [HTMLNode::Element {
+            name: "html".into(),
+            attrs: [].into(),
+            children: [
+                HTMLNode::Element {
+                    name: "head".into(),
+                    attrs: [].into(),
+                    children: [].into(),
+                },
+                HTMLNode::Element {
+                    name: "body".into(),
+                    attrs: [].into(),
+                    children: [HTMLNode::Element {
+                        name: "a".into(),
+                        attrs: [].into(),
+                        children: [HTMLNode::Text(" ".into())].into(),
+                    }]
+                    .into(),
+                }
+            ]
+            .into()
+        }]);
+
+        let soup = Soup::html(
+            r#"
+<!DOCTYPE html>
+<html lang="en">
+
+    <head>
+        <meta charset="UTF-8"/>
+        <title>Hello!</title>
+    </head>
+
+    <body>
+        <a>Link</a>
+        <span> </span>
+    </body>
+</html>"#,
+        );
+        assert_eq!(soup.nodes, [
+            HTMLNode::Doctype("html".into()),
+            HTMLNode::Element {
+                name: "html".into(),
+                attrs: [("lang".into(), "en".into())].into(),
+                children: [
+                    HTMLNode::Element {
+                        name: "head".into(),
+                        attrs: [].into(),
+                        children: [
+                            HTMLNode::Text("\n        ".into()),
+                            HTMLNode::Void {
+                                name: "meta".into(),
+                                attrs: [("charset".into(), "UTF-8".into())].into()
+                            },
+                            HTMLNode::Text("\n        ".into()),
+                            HTMLNode::Element {
+                                name: "title".into(),
+                                attrs: [].into(),
+                                children: [HTMLNode::Text("Hello!".into())].into()
+                            },
+                            HTMLNode::Text("\n    ".into()),
+                        ]
+                        .into(),
+                    },
+                    HTMLNode::Text("\n\n    ".into()),
+                    HTMLNode::Element {
+                        name: "body".into(),
+                        attrs: [].into(),
+                        children: [
+                            HTMLNode::Text("\n        ".into()),
+                            HTMLNode::Element {
+                                name: "a".into(),
+                                attrs: [].into(),
+                                children: [HTMLNode::Text("Link".into())].into(),
+                            },
+                            HTMLNode::Text("\n        ".into()),
+                            HTMLNode::Element {
+                                name: "span".into(),
+                                attrs: [].into(),
+                                children: [HTMLNode::Text(" ".into())].into(),
+                            },
+                            HTMLNode::Text("\n    \n".into())
+                        ]
+                        .into(),
+                    }
+                ]
+                .into()
+            }
+        ]);
     }
 }

--- a/src/parser/html/mod.rs
+++ b/src/parser/html/mod.rs
@@ -1,13 +1,13 @@
 #[cfg(feature = "html-lenient")]
 mod lenient;
-#[cfg(feature = "html")]
+#[cfg(any(feature = "html-lenient", feature = "html-strict"))]
 mod node;
 #[cfg(feature = "html-strict")]
 mod strict;
 
 #[cfg(feature = "html-lenient")]
 pub use lenient::LenientHTMLParser;
-#[cfg(feature = "html")]
+#[cfg(any(feature = "html-lenient", feature = "html-strict"))]
 pub use node::HTMLNode;
 #[cfg(feature = "html-strict")]
 pub use strict::StrictHTMLParser;

--- a/src/parser/html/node.rs
+++ b/src/parser/html/node.rs
@@ -145,6 +145,8 @@ mod tests {
 
         assert_eq!(nodes.next().unwrap().name(), Some(&"body"));
 
+        assert_eq!(nodes.next().unwrap(), &HTMLNode::Text("\n        "));
+
         assert_eq!(nodes.next().unwrap(), &HTMLNode::Element {
             name: "h1",
             attrs: BTreeMap::default(),
@@ -152,6 +154,8 @@ mod tests {
         });
 
         assert_eq!(nodes.next().unwrap(), &HTMLNode::Text("Hello World!"));
+
+        assert_eq!(nodes.next().unwrap(), &HTMLNode::Text("\n        "));
 
         assert_eq!(nodes.next().unwrap(), &HTMLNode::Element {
             name: "p",
@@ -173,11 +177,15 @@ mod tests {
 
         let mut nodes = body.into_iter();
 
+        assert_eq!(nodes.next().unwrap(), &HTMLNode::Text("\n        "));
+
         assert_eq!(nodes.next().unwrap(), &HTMLNode::Element {
             name: "h1",
             attrs: BTreeMap::default(),
             children: vec![HTMLNode::Text("Hello World!")]
         });
+
+        assert_eq!(nodes.next().unwrap(), &HTMLNode::Text("\n        "));
 
         assert_eq!(nodes.next().unwrap(), &HTMLNode::Element {
             name: "p",

--- a/src/parser/html/strict.rs
+++ b/src/parser/html/strict.rs
@@ -1,3 +1,5 @@
+#![allow(clippy::type_complexity)]
+
 use std::marker::PhantomData;
 
 use nom::{
@@ -200,7 +202,7 @@ fn element(i: &str) -> IResult<&str, HTMLNode<&str>> {
 }
 
 fn text(i: &str) -> IResult<&str, HTMLNode<&str>> {
-    map(map(is_not("<"), str::trim), HTMLNode::Text)(i)
+    map(is_not("<"), HTMLNode::Text)(i)
 }
 
 fn single(i: &str) -> IResult<&str, HTMLNode<&str>> {
@@ -208,7 +210,7 @@ fn single(i: &str) -> IResult<&str, HTMLNode<&str>> {
 }
 
 pub(crate) fn parse(i: &str) -> IResult<&str, Vec<HTMLNode<&str>>> {
-    many0(ws(single))(i)
+    many0(single)(i)
 }
 
 #[allow(clippy::too_many_lines)]
@@ -395,6 +397,14 @@ mod test {
             }))
         );
         assert_eq!(
+            element("<a> </a>"),
+            Ok(("", HTMLNode::Element {
+                name: "a",
+                attrs: [].into(),
+                children: [HTMLNode::Text(" ")].into()
+            }))
+        );
+        assert_eq!(
             element(r#"<a rel=""></a>"#),
             Ok(("", HTMLNode::Element {
                 name: "a",
@@ -434,28 +444,30 @@ mod test {
         assert_eq!(
             parse("\t\t<!-- Hello -->\n\t<!doctype html>\n<!-- second -->"),
             Ok(("", vec![
+                HTMLNode::Text("\t\t"),
                 HTMLNode::Comment(" Hello "),
+                HTMLNode::Text("\n\t"),
                 HTMLNode::Doctype("html"),
+                HTMLNode::Text("\n"),
                 HTMLNode::Comment(" second ")
             ]))
         );
 
         assert_eq!(
             parse(
-                r#"
-                <!--Here's a link.-->
+                r#"<!--Here's a link.-->
                 <a href="https://example.com"/>
-                With some text.
-            "#
+                With some text."#
             ),
             Ok(("", vec![
                 HTMLNode::Comment("Here's a link."),
+                HTMLNode::Text("\n                "),
                 HTMLNode::Element {
                     name: "a",
                     attrs: [("href", "https://example.com")].into(),
                     children: [].into()
                 },
-                HTMLNode::Text("With some text.")
+                HTMLNode::Text("\n                With some text.")
             ])),
         );
 
@@ -469,25 +481,36 @@ mod test {
                 </div>
             "#
             ),
-            Ok(("", vec![HTMLNode::Element {
-                name: "div",
-                attrs: [("class", "outer")].into(),
-                children: vec![HTMLNode::Element {
+            Ok(("", vec![
+                HTMLNode::Text("\n                "),
+                HTMLNode::Element {
                     name: "div",
-                    attrs: [("class", "inner")].into(),
-                    children: vec![HTMLNode::Element {
-                        name: "p",
-                        attrs: [].into(),
-                        children: vec![HTMLNode::Text("Hello, world!")],
-                    }],
-                }],
-            }])),
+                    attrs: [("class", "outer")].into(),
+                    children: vec![
+                        HTMLNode::Text("\n                    "),
+                        HTMLNode::Element {
+                            name: "div",
+                            attrs: [("class", "inner")].into(),
+                            children: vec![
+                                HTMLNode::Text("\n                        "),
+                                HTMLNode::Element {
+                                    name: "p",
+                                    attrs: [].into(),
+                                    children: vec![HTMLNode::Text("Hello, world!")],
+                                },
+                                HTMLNode::Text("\n                    "),
+                            ],
+                        },
+                        HTMLNode::Text("\n                "),
+                    ],
+                },
+                HTMLNode::Text("\n            "),
+            ])),
         );
 
         assert_eq!(
             parse(
-                r#"
-<script type="application/javascript">
+                r#"<script type="application/javascript">
 if (1 < 2) {
     console.log("Hello, world!");
 }
@@ -510,38 +533,49 @@ if (1 < 2) {
                     attrs: [("type", "application/javascript")].into(),
                     content: "if (1 < 2) {\n    console.log(\"Hello, world!\");\n}",
                 },
+                HTMLNode::Text("\n"),
                 HTMLNode::Element {
                     name: "div",
                     attrs: [("class", "outer")].into(),
                     children: vec![
+                        HTMLNode::Text("\n    "),
                         HTMLNode::Element {
                             name: "div",
                             attrs: [("class", "inner")].into(),
                             children: vec![
+                                HTMLNode::Text("\n        "),
                                 HTMLNode::Element {
                                     name: "p",
                                     attrs: [].into(),
                                     children: vec![HTMLNode::Text("Hello, world!")],
                                 },
+                                HTMLNode::Text("\n        "),
                                 HTMLNode::Element {
                                     name: "p",
                                     attrs: [].into(),
                                     children: vec![HTMLNode::Text("Another element...")],
                                 },
-                                HTMLNode::Text("Just some text...")
+                                HTMLNode::Text("\n        Just some text...\n    ")
                             ],
                         },
+                        HTMLNode::Text("\n    "),
                         HTMLNode::Element {
                             name: "div",
                             attrs: [].into(),
-                            children: vec![HTMLNode::Element {
-                                name: "p",
-                                attrs: [].into(),
-                                children: vec![HTMLNode::Text("Fancy nesting")],
-                            }]
-                        }
+                            children: vec![
+                                HTMLNode::Text("\n        "),
+                                HTMLNode::Element {
+                                    name: "p",
+                                    attrs: [].into(),
+                                    children: vec![HTMLNode::Text("Fancy nesting")],
+                                },
+                                HTMLNode::Text("\n    "),
+                            ]
+                        },
+                        HTMLNode::Text("\n"),
                     ],
-                }
+                },
+                HTMLNode::Text("\n"),
             ])),
         );
     }

--- a/src/parser/mod.rs
+++ b/src/parser/mod.rs
@@ -1,7 +1,9 @@
+#[cfg(any(feature = "html-lenient", feature = "html-strict"))]
 mod html;
 #[cfg(feature = "xml")]
 mod xml;
 
+#[cfg(any(feature = "html-lenient", feature = "html-strict"))]
 pub use html::*;
 #[cfg(feature = "xml")]
 pub use xml::*;

--- a/src/query.rs
+++ b/src/query.rs
@@ -19,9 +19,9 @@ pub struct Query<'x, N, F> {
     filter: F,
 }
 
-impl<'x, N, F> Copy for Query<'x, N, F> where F: Copy {}
+impl<N, F> Copy for Query<'_, N, F> where F: Copy {}
 
-impl<'x, N, F> Clone for Query<'x, N, F>
+impl<N, F> Clone for Query<'_, N, F>
 where
     F: Clone,
 {
@@ -273,7 +273,7 @@ pub struct QueryItem<'x, N> {
     item: &'x N,
 }
 
-impl<'x, N> QueryItem<'x, N>
+impl<N> QueryItem<'_, N>
 where
     N: Node + Clone,
 {
@@ -286,7 +286,7 @@ where
     }
 }
 
-impl<'x, N> std::ops::Deref for QueryItem<'x, N> {
+impl<N> std::ops::Deref for QueryItem<'_, N> {
     type Target = N;
 
     fn deref(&self) -> &Self::Target {

